lib/services/authentication.go

```diff
@@ -19,13 +19,401 @@ limitations under the License.
 package services
 
 import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/x509"
 	"encoding/json"
 	"fmt"
+	"time"
 
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/utils"
+
 	"github.com/gravitational/trace"
+	"github.com/pquerna/otp/totp"
+	"github.com/tstranex/u2f"
+	"golang.org/x/crypto/bcrypt"
 )
 
+// AuthPreference defines the authentication preferences for a specific
+// cluster. It defines the type (local, oidc) and second factor (off, otp, oidc).
+// AuthPreference is a configuration resource, never create more than one instance
+// of it.
+type AuthPreference interface {
+	// Resource provides common resource properties.
+	Resource
+
+	// GetType gets the type of authentication: local, saml, or oidc.
+	GetType() string
+	// SetType sets the type of authentication: local, saml, or oidc.
+	SetType(string)
+
+	// GetSecondFactor gets the type of second factor: off, otp or u2f.
+	GetSecondFactor() string
+	// SetSecondFactor sets the type of second factor: off, otp, or u2f.
+	SetSecondFactor(string)
+
+	// GetConnectorName gets the name of the OIDC or SAML connector to use. If
+	// this value is empty, we fall back to the first connector in the backend.
+	GetConnectorName() string
+	// SetConnectorName sets the name of the OIDC or SAML connector to use. If
+	// this value is empty, we fall back to the first connector in the backend.
+	SetConnectorName(string)
+
+	// GetU2F gets the U2F configuration settings.
+	GetU2F() (*U2F, error)
+	// SetU2F sets the U2F configuration settings.
+	SetU2F(*U2F)
+
+	// CheckAndSetDefaults sets and default values and then
+	// verifies the constraints for AuthPreference.
+	CheckAndSetDefaults() error
+
+	// String represents a human readable version of authentication settings.
+	String() string
+}
+
+// NewAuthPreference is a convenience method to to create AuthPreferenceV2.
```

Comment:
nit: when the implementation changes to instead construct a value of type with the next version, this comment will be obsolete.
```suggestion
// NewAuthPreference is a convenience method to to create an auth preference value
```